src/config/source/yaml_file.cpp

```diff
@@ -0,0 +1,271 @@
+   std::ifstream file(path, std::ios::in | std::ios::binary);
+   if (file.fail()) {
+      throw std::runtime_error(fmt::format("Could not open the YAML file: '{}'", path));
+   }
+
+   std::ostringstream contents;
+   if (file.peek() != std::ifstream::traits_type::eof()) {
+      contents << file.rdbuf();
+   }
+   if (contents.fail()) {
+      throw std::runtime_error(fmt::format("Error when reading the YAML file: '{}'", path));
+   }
+
+   return fromYAML(fmt::format("file: '{}'", path.string()), contents.str());
+}
+
+std::string YamlFile::debugContext() const {
+   return fmt::format("YAML file '{}'", debug_context);
+}
+
+namespace {
+ConfigValue yamlNodeToConfigValue(
+   const ConfigAttributeSpecification& attribute_spec,
+   const YAML::Node& yaml
+) {
+   try {
+      switch (attribute_spec.type) {
+         case ConfigValueType::STRING:
+            return ConfigValue::fromString(yaml.as<std::string>());
+         case ConfigValueType::PATH:
+            return ConfigValue::fromPath({std::filesystem::path{yaml.as<std::string>()}});
+         case ConfigValueType::INT32:
+            return ConfigValue::fromInt32(yaml.as<int32_t>());
+         case ConfigValueType::UINT32:
+            return ConfigValue::fromUint32(yaml.as<uint32_t>());
+         case ConfigValueType::UINT16:
+            return ConfigValue::fromUint16(yaml.as<uint16_t>());
+         case ConfigValueType::BOOL:
+            return ConfigValue::fromBool(yaml.as<bool>());
+      }
+      SILO_UNREACHABLE();
+   } catch (YAML::BadConversion& error) {
+      throw ConfigException(fmt::format(
+         "cannot parse '{}' as {}: {}",
+         yaml,
+         configValueTypeToString(attribute_spec.type),
+         error.what()
+      ));
+   }
+}
+}  // namespace
+
+VerifiedConfigAttributes YamlFile::verify(const ConfigSpecification& config_specification) const {
+   // No need to stringify and do duplicate check since
+   // ConfigKeyPath is actually directly representing YAML paths.
+
+   // Check the ones given, collect erroneous ones in foo.bar syntax
+   std::vector<std::string> invalid_config_keys;
+   std::unordered_map<ConfigKeyPath, ConfigValue> provided_config_values;
+   for (const auto& [key, yaml] : getYamlFields()) {
+      auto attribute_spec = config_specification.getAttributeSpecification(key);
+      if (!attribute_spec.has_value()) {
+         invalid_config_keys.push_back(configKeyPathToString(key));
+      } else {
+         const ConfigValue value = yamlNodeToConfigValue(attribute_spec.value(), yaml);
+         provided_config_values.emplace(key, value);
+      }
+   }
+
+   if (!invalid_config_keys.empty()) {
+      const char* keys_or_options = (invalid_config_keys.size() >= 2) ? "keys" : "key";
+      throw silo::config::ConfigException(fmt::format(
+         "in {}: unknown {} {}",
+         debugContext(),
+         keys_or_options,
+         boost::join(invalid_config_keys, ", ")
+      ));
+   }
```

Comment:
This probably makes sense (and could be enabled just for the YAML files, and via a command line or env var option). But can we do this in a separate PR? Alex and I specified that we will report errors on unknown keys as part of this PR, so this comes late. It's probably "not much work" but we want to submit PRs as soon as possible, so this doesn't make it.

Reply:
IMO this is a blocker - it will break the current setup of our GenSpectrum deployments. Adapting to those new changes will make it more complicated. 

I believe the same is true for Loculus.

Reply:
The DatabaseConfig is not using this new interface. From the view-point of SILO I would actually like to rename it to just DatabaseSchema instead

Reply:
Ah good point - indeed, this doesn't overlap with LAPIS config :+1: